### PR TITLE
Feature: New Webhook Triggers, KV Store, and Shopify integration

### PR DESCRIPTION
**Pull request by nicktrn**
Closes #<issue>

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/triggerdotdev/trigger.dev/blob/main/CONTRIBUTING.md)
- [x] The PR title follows the convention.
- [x] I ran and tested the code works

---

## Todo

- [x] Finish docs
- [x] Fix payload types.. again
- [x] More testing

---

## Testing

KV Store

https://github.com/triggerdotdev/trigger.dev/blob/813ea1e17f60612f6a98c77a74f931e52f784b2d/references/job-catalog/src/built-ins.ts#L302-L328

Shopify tasks

https://github.com/triggerdotdev/trigger.dev/blob/813ea1e17f60612f6a98c77a74f931e52f784b2d/references/job-catalog/src/shopify.ts#L45-L95

---

## Changelog

- Webhook Triggers using HTTP Endpoint
- ~~Webhook deliveries are now jobs~~
- Deliveries are shown in the UI
- Key-Value Store
- A few fixes here and there
- Shopify integration

---

## Screenshots

_[Screenshots]_

💯

### CODE REVIEW

File: apps/webapp/app/routes/api.v1.store.$action.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    return json({ error: "Invalid request body" }, { status: 400 });
+  }
+
+  const store = new KeyValueStore(authenticatedEnv);
+
+  const { action, key } = body.data;
+
+  try {
+    switch (action) {
+      case "GET": {
```
Comment (ericallam) [inline apps/webapp/app/routes/api.v1.store.$action.ts]: This should just use an HTTP verb instead I think. So doing a GET to /api/v1/store/$key to GET a key's value, doing a PUT to /api/v1/store/$key to create/update a value, and doing a DELETE to /api/v1/store/$key to delete it.

File: apps/webapp/app/routes/api.v1.store.$action.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  const authenticatedEnv = authenticationResult.environment;
+
+  // Now parse the request body
+  const anyBody = await request.json();
+
+  const bodyObject = z.object({}).passthrough().parse(anyBody);
+
+  const body = KeyValueStoreRequestBodySchema.safeParse({
```
Comment (ericallam) [inline apps/webapp/app/routes/api.v1.store.$action.ts]: Let's make sure we limit the size of the value to something like 256KB, so this isn't used as a DB alternative.

**Review by ericallam — DISMISSED** [review]

File: apps/webapp/app/routes/api.v1.store.$action.ts (hunk 1 above)
Comment (nicktrn) [reply apps/webapp/app/routes/api.v1.store.$action.ts]: Thought you may flag this - let's do it right.

File: apps/webapp/app/routes/_app.orgs.$organizationSlug.projects.$projectParam.triggers_.webhooks.$triggerParam._index/route.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        of attempted registrations below.
+      </Paragraph>
+
+      {!trigger.active &&
+        <Form method="post" {...form.props}>
+        <Callout variant="error" className="justiy-between mb-4 items-center">
+          <Paragraph variant="small" className={cn(variantClasses.error.textColor, "grow")}>
+            Registration hasn't succeeded yet, check the runs below.
+          </Paragraph>
+          {/* <input
```
Comment (nicktrn) [inline apps/webapp/app/routes/_app.orgs.$organizationSlug.projects.$projectParam.triggers_.webhooks.$triggerParam._index/route.tsx:148]: Removed the Registration Retry button as this page shows Registration Runs across all Environments. Might need a UI tweak to display the Retry button on failed Run rows instead. (Will have to make minor changes to the backend too.)

File: apps/webapp/app/routes/api.v1.webhooks.$key.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import { authenticateApiRequest } from "~/services/apiAuth.server";
+import { logger } from "~/services/logger.server";
+import { UpdateWebhookService } from "~/services/sources/updateWebhook.server";
+
+const ParamsSchema = z.object({
+  key: z.string(),
+});
+
+export async function action({ request, params }: ActionFunctionArgs) {
+  logger.info("Updating webhook", { url: request.url });
```
Comment (nicktrn) [inline apps/webapp/app/routes/api.v1.webhooks.$key.ts:14]: This route exists to mark webhooks as active and update the config

File: apps/webapp/app/services/sources/handleWebhookRequest.server.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    if (!webhookEnvironment.active) {
+      return { status: 200 };
+    }
+
+    const webhookRequest = await createHttpSourceRequest(request);
+
+    const lockId = webhookIdToLockId(webhookEnvironment.webhookId);
+
+    await this.#prismaClient.$transaction(async (tx) => {
+      await tx.$executeRaw`SELECT pg_advisory_xact_lock(${lockId})`;
```
Comment (nicktrn) [inline apps/webapp/app/services/sources/handleWebhookRequest.server.ts:40]: :copyright: @ericallam

File: apps/webapp/app/services/sources/updateWebhook.server.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    await this.#prismaClient.webhook.update({
+      where: {
+        key_projectId: {
+          key,
+          projectId: environment.projectId,
+        },
+      },
+      data: {
+        active: payload.active,
```
Comment (nicktrn) [inline apps/webapp/app/services/sources/updateWebhook.server.ts:39]: This means the webhook is active across all enabled environments, but may want a better check for this.

File: apps/webapp/app/services/triggers/registerWebhook.server.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    if (!upsertResult) {
+      return;
+    }
+
+    const { webhook, webhookEnvironment } = upsertResult;
+    const { config, desiredConfig } = webhookEnvironment;
+
+    if (webhook.active && isEqual(config, desiredConfig)) {
+      return;
```
Comment (nicktrn) [inline apps/webapp/app/services/triggers/registerWebhook.server.ts:48]: Don't trigger registration runs if config unchanged

File: integrations/shopify/src/rest.ts [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+export class Rest {
+  constructor(
+    private runTask: ShopifyRunTask,
+    private session: Session
+  ) {}
+}
+
+interface MergeProxyConstructor {
```
Comment (nicktrn) [inline integrations/shopify/src/rest.ts:293]: This is to have correct types if we decide to extend the Rest class, i.e. add more tasks

File: integrations/shopify/src/rest.ts [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+  new <TTarget extends Record<any, any>, TResult extends Record<any, any>>(
+    target: TTarget,
+    handler: ProxyHandler<TResult>
+  ): TTarget & TResult;
+}
+
+type ResourceMap = {
+  [KResourceType in ResourcesWithStandardMethods]: Resource<KResourceType>;
```
Comment (nicktrn) [inline integrations/shopify/src/rest.ts:301]: Only a subset of Shopify Resources expose all standard methods. Need the raw client for ones that don't.

File: integrations/shopify/src/webhooks.ts [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    schemas: {
+      params: z.object({
+        topic: WebhookTopicSchema,
+        fields: z.string().array().optional(),
+      }),
+      // config: z.record(z.string().array()),
+    },
+    version: "0.1.0",
+    integration,
+    key: (params) => `${params.topic}-${params.fields?.join(".")}`,
```
Comment (nicktrn) [inline integrations/shopify/src/webhooks.ts]: This key will get hashed

File: integrations/shopify/src/webhooks.ts [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    crud: {
+      create: async ({ io, ctx }) => {
+        const webhook = await io.integration.rest.Webhook.save("create-webhook", {
+          fromData: {
+            address: ctx.url,
+            topic: ctx.params.topic,
+            fields: ctx.params.fields,
+            // fields: ctx.config.desired.fields,
```
Comment (nicktrn) [inline integrations/shopify/src/webhooks.ts]: `fields` is an optional array of property keys the payload should include. Includes all if not specified. Up for debate if this should be a param or config that may be merged across triggers. It shouldn't matter too much either way.

File: packages/trigger-sdk/src/apiClient.ts [hunk 12]
```diff
@@ -716,6 +825,29 @@ async function zodfetchWithVersions<
   };
 }
 
+async function fetchHead(
```
Comment (nicktrn) [inline packages/trigger-sdk/src/apiClient.ts:828]: Just a HEAD request with retries.

File: packages/trigger-sdk/src/triggerClient.ts [hunk 13]
```diff
@@ -67,8 +75,39 @@ import type {
...
+  });
+};
+
+const deliverWebhookEvent = (key: string): EventSpecification<Request> => ({
+  name: `${DELIVER_WEBHOOK_REQUEST}.${key}`,
+  title: "Deliver Webhook",
+  source: "internal",
+  icon: "mail-fast",
+  parsePayload: parseRequestPayload,
+});
```
Comment (nicktrn) [inline packages/trigger-sdk/src/triggerClient.ts:101]: Now unused, can remove or comment out.

File: packages/trigger-sdk/src/triggers/eventTrigger.ts [hunk 14]
```diff
@@ -1,15 +1,23 @@
...
 import { formatSchemaErrors } from "../utils/formatSchemaErrors";
 import { ParsedPayloadSchemaError } from "../errors";
+import { VerifyCallback } from "../httpEndpoint";
 
 type EventTriggerOptions<TEventSpecification extends EventSpecification<any>> = {
   event: TEventSpecification;
   name?: string | string[];
   source?: string;
   filter?: EventFilter;
+  verify?: EventTypeFromSpecification<TEventSpecification> extends Request ? VerifyCallback : never;
```
Comment (nicktrn) [inline packages/trigger-sdk/src/triggers/eventTrigger.ts:20]: Don't recall why I added this, but it made sense at the time

File: apps/webapp/app/components/runs/WebhookDeliveryRunsTable.tsx [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+          </TableBlankRow>
+        ) : runs.length === 0 ? (
+          <TableBlankRow colSpan={8}>
+            <NoRuns title="No Runs match your filters" />
+          </TableBlankRow>
+        ) : (
+          runs.map((run) => {
+            const path = `.`;
+
+            const doNothing = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
```
Comment (matt-aitken) [inline apps/webapp/app/components/runs/WebhookDeliveryRunsTable.tsx]: This seems a bit strange, if you removed the `onClick` from the TableCells was it formatting weirdly?

File: apps/webapp/app/routes/_app.orgs.$organizationSlug.projects.$projectParam.triggers_.webhooks.$triggerParam.delivery/DeliveryListPagination.tsx [hunk 16]
```diff
@@ -0,0 +1,56 @@
+import { useLocation } from "@remix-run/react";
+import { LinkButton } from "~/components/primitives/Buttons";
+import { Direction, RunList } from "~/presenters/RunListPresenter.server";
+import { cn } from "~/utils/cn";
+
+export function DeliveryListPagination({ list, className }: { list: RunList; className?: string }) {
```
Comment (matt-aitken) [inline apps/webapp/app/routes/_app.orgs.$organizationSlug.projects.$projectParam.triggers_.webhooks.$triggerParam.delivery/DeliveryListPagination.tsx]: Couldn't some code be shared between this and the existing RunListPagination? Should we just use these same component for all pagination and call them just ListPagination?

File: packages/database/prisma/migrations/20231115134828_add_events_schema_and_tables/migration.sql [hunk 17]
```diff
@@ -1,5 +1,7 @@
 CREATE SCHEMA IF NOT EXISTS triggerdotdev_events;
 
+DROP TABLE IF EXISTS triggerdotdev_events.run_executions;
+
```
Comment (matt-aitken) [inline packages/database/prisma/migrations/20231115134828_add_events_schema_and_tables/migration.sql:4]: Should this be in here?

**Review by matt-aitken — CHANGES_REQUESTED** [review]

**Review by matt-aitken — APPROVED** [review]

**Pull request merged**